app/controllers/forest_liana/smart_actions_controller.rb

```diff
@@ -8,39 +12,70 @@ class SmartActionsController < ForestLiana::ApplicationController
 
     private
 
+    def smart_action_pre_perform_checks
+      get_smart_action_request
+      find_resource
+      check_permission_for_smart_route
+      ensure_record_ids_in_scope
+    end
+
     def get_smart_action_request
       begin
         params[:data][:attributes]
+        @parameters = ForestLiana::Ability::Permission::RequestPermission::decodeSignedApprovalRequest(params.permit!)
       rescue => error
         FOREST_REPORTER.report error
         FOREST_LOGGER.error "Smart Action execution error: #{error}"
         {}
       end
     end
 
-    def smart_action_pre_perform_checks
-      check_permission_for_smart_route
-      ensure_record_ids_in_scope
+    def find_resource
+      begin
+        @resource = SchemaUtils.find_model_from_collection_name(@parameters[:data][:attributes][:collection_name])
+        if @resource.nil? || !SchemaUtils.model_included?(@resource) ||
+          !@resource.ancestors.include?(ActiveRecord::Base)
+          render serializer: nil, json: { status: 404 }, status: :not_found
+        end
+        @resource
+      rescue => error
+        FOREST_REPORTER.report error
+        FOREST_LOGGER.error "Find Collection error: #{error}\n#{format_stacktrace(error)}"
+        render serializer: nil, json: { status: 404 }, status: :not_found
+      end
+    end
+
+    def check_permission_for_smart_route
+      smart_action_request = @parameters[:data][:attributes]
+      if !smart_action_request.nil? && smart_action_request.has_key?(:smart_action_id)
+        forest_authorize!(
+          'action',
+          forest_user,
+          @resource,
+          {parameters: params, endpoint: request.fullpath.split('?').first, http_method: request.request_method}
```

Comment:
Okay. You pass the input params so forest_authorize re-decode (decodeSignedApprovalRequest). Looks pretty good to me.